Script Modules API: Move the script modules to the footer in classic themes

luisherranz opened the pull request:
Trac ticket: https://core.trac.wordpress.org/ticket/60240

---

## What

Moves the script module prints to the footer in classic themes. It also removes the ability to call the print functions more than once because it's now unnecessary.

## Why

In the ticket that introduced the Script Modules API ([#56313](https://core.trac.wordpress.org/ticket/56313 "#56313: enhancement: Add the ability to handle ES Modules and Import Maps (reopened)")), I added a logic that prints as many enqueued and preloads as possible in the head, then prints the remaining ones in the footer and includes the import map. I explain my reasoning [​in this comment](https://github.com/WordPress/wordpress-develop/pull/5818#discussion_r1446384970).

But [@cbravobernal](https://profiles.wordpress.org/cbravobernal) realized that [import maps fail if they appear after the modules](https://github.com/WordPress/gutenberg/pull/57778#discussion_r1449482982). That means that there's no other way but printing the import map first, and then all the modules and preloads (yes, preloads also fail!). So we need to check if we are on a classic or block theme, and if we are in a classic theme, move everything to the footer.

## How

By leveraging `wp_is_block_theme()` and moving everything to the footer in classic themes:

```php
public function add_hooks() {
  $position = wp_is_block_theme() ? 'wp_head' : 'wp_footer';
  add_action( $position, array( $this, 'print_import_map' ) );
  add_action( $position, array( $this, 'print_enqueued_modules' ) );
  add_action( $position, array( $this, 'print_module_preloads' ) );
}
```

## Additional comments

@westonruter mentioned that we could optimize this in the future if we have a way to filter the output before sending the response to the client:

- https://github.com/WordPress/gutenberg/pull/57778#discussion_r1450806250

## Testing instructions

To check that this fixes the order of the import map:

1. Create a new plugin with three files:
   ```php
   // test-plugin/test.php
   <?php
   /*
   Plugin Name: Test Script Modules
   Version: 1.0.0
   */

   wp_register_script_module(
     'bar',
     plugins_url( '/bar.js', __FILE__ )
   );

   wp_enqueue_script_module(
     'foo',
     plugins_url( '/foo.js', __FILE__ ),
     array( 'bar' )
   );
   ```
   ```js
   // test-plugin/foo.js
   import bar from "bar";

   bar();
   ```
   ```js
   // test-plugin/bar.js
   export default function bar() {
     console.log("bar");
   }
   ```
2. Activate the plugin.
3. Open the site (frontend).
4. Check that `"bar"` was printed in the console.

To check that this fixes the positioning of the scripts/link in the classic themes:

1. Load a block theme.
2. Check that the scripts with `type="importmap"` and `type="module"`, and the link with `rel="modulepreload"` are printed in the head.
3. Load a classic theme.
4. Check that the scripts with `type="importmap"` and `type="module"`, and the link with `rel="modulepreload"` are printed in the footer.

luisherranz committed:
Move all hooks to the footer in classic themes and remove multiple prints functionality

sarinajor commented:
i am also facing same problem on my 2 sites
https://eventmanagementexpert.com/
https://eventplannertoronto.ca/

luisherranz committed:
Fix print functions missing script in their names

luisherranz commented:
FWIW, as I removed the ability to call the print functions more than once because it's now unnecessary, I also removed the `test_print_enqueued_script_modules_can_be_called_multiple_times` and `test_print_preloaded_script_modules_can_be_called_multiple_times` tests.

youknowriad commented:
Committed in https://core.trac.wordpress.org/changeset/57345

cbravobernal commented:
I tested and is working as expected.

Trunk screenshots:
![Screenshot 2024-01-24 at 11 42 32](https://github.com/WordPress/wordpress-develop/assets/37012961/4ae1bddb-ae96-4aff-80f8-6c92282fd629)
![Screenshot 2024-01-24 at 11 42 51](https://github.com/WordPress/wordpress-develop/assets/37012961/dc7d07f6-8659-417a-9d7e-9f0c6840eb51)
 - Is bar in console log ❌
 - There are errors on console ❌
 -  The scripts are in the footer in a classic theme. ❌

PR applied screenshots:
![Screenshot 2024-01-24 at 11 46 18](https://github.com/WordPress/wordpress-develop/assets/37012961/cb6d89c3-ced6-496f-b392-c09218e32a7c)
![Screenshot 2024-01-24 at 11 46 12](https://github.com/WordPress/wordpress-develop/assets/37012961/8a8a2d34-3e07-4053-ae9a-627011cc5f48)

 - Is bar in console log ✅
 - There are no errors on console ✅
 -  The scripts are in the footer in a classic theme. ✅

luisherranz commented:
Thanks, folks!
